File: portable/test_conformance.py
"""Canonical Python conformance test for portable_sar_verify.py against the
shared synthetic fixture corpus (the SAME JSON files the JS test consumes —
cross-language parity is the point). Run with: python3 test_conformance.py
"""
import base64
import json
import os
import sys

sys.path.insert(0, os.path.dirname(__file__))
from portable_sar_verify import (  # noqa: E402
    KeyPolicyEntry,
    PORTABLE_PROFILE_ID,
    STATUS_VERIFIED,
    STATUS_UNKNOWN_KEY,
    STATUS_PROFILE_NOT_AUTHORIZED,
    STATUS_RECEIPT_ID_MISMATCH,
    STATUS_INVALID_SIGNATURE,
    STATUS_NOT_CANDIDATE,
    verify_portable_sar_receipt,
)

HERE = os.path.dirname(__file__)
fixtures = json.load(open(os.path.join(HERE, "fixtures", "portable-sar-fixtures.json")))
key_registry = json.load(open(os.path.join(HERE, "fixtures", "portable-sar-fixture-keys.json")))


def b64url_to_bytes(s: str) -> bytes:
    pad = (-len(s)) % 4
    return base64.urlsafe_b64decode(s + ("=" * pad))


def key_policy(kid):
    entry = key_registry.get(kid)
    if entry is None:
        return None
    return KeyPolicyEntry(
        pubkey=b64url_to_bytes(entry["pubkey_b64url"]),
        profiles=frozenset(entry["profiles"]),
        source=entry["source"],
    )


EXPECTED = {
    "01_valid_portable": STATUS_VERIFIED,
    "02_unsigned_counterparty": STATUS_VERIFIED,
    "03_with_ext": STATUS_VERIFIED,
    "04_tampered_signed_field": STATUS_RECEIPT_ID_MISMATCH,
    "05_substituted_unsigned_metadata": STATUS_VERIFIED,
    "05b_stripped_unsigned_metadata": STATUS_VERIFIED,
    "06_unknown_key": STATUS_UNKNOWN_KEY,
    "07_spoofed_key": STATUS_INVALID_SIGNATURE,
    "08_invalid_signature": STATUS_INVALID_SIGNATURE,
    "09_malformed_version": STATUS_NOT_CANDIDATE,
    "09b_malformed_alg": STATUS_NOT_CANDIDATE,
    "11_walletbound_presented_as_portable": STATUS_NOT_CANDIDATE,
    "12_metadata_grafted_walletbound": STATUS_PROFILE_NOT_AUTHORIZED,
    "13_portable_presented_as_walletbound_boundary_case": STATUS_VERIFIED,
    "14_ambiguous_envelope": STATUS_VERIFIED,
}

passed = 0
failed = 0

for name, expected_status in EXPECTED.items():
    receipt = fixtures[name]
    result = verify_portable_sar_receipt(receipt, key_policy)
    ok = result.status == expected_status
    passed += ok
    failed += not ok
    print(f"{'ok  ' if ok else 'FAIL'}  {name:<45} expected={expected_status} got={result.status}" + ("" if ok else f"  reason={result.reason}"))

for name, expected_status in EXPECTED.items():
    if expected_status != STATUS_VERIFIED:
        continue
    result = verify_portable_sar_receipt(fixtures[name], key_policy)
    ok = result.wallet_binding_attested is False
    passed += ok
    failed += not ok
    print(f"{'ok  ' if ok else 'FAIL'}  {name:<45} wallet_binding_attested must be False, got={result.wallet_binding_attested}")

r = verify_portable_sar_receipt(fixtures["12_metadata_grafted_walletbound"], key_policy, requested_profile=PORTABLE_PROFILE_ID)
ok = r.status == STATUS_PROFILE_NOT_AUTHORIZED
passed += ok
failed += not ok
print(f"{'ok  ' if ok else 'FAIL'}  explicit_request_cannot_override_authorization got={r.status}")

print(f"\n{passed} passed, {failed} failed")
sys.exit(0 if failed == 0 else 1)

File: portable/test_conformance.mjs
// Canonical JS conformance test for portable_sar_verify.js against the
// shared synthetic fixture corpus. Run with: node test_conformance.mjs
import { verifyPortableSarReceipt, STATUS } from "./portable_sar_verify.js";
import fs from "fs";

const fixtures = JSON.parse(fs.readFileSync(new URL("./fixtures/portable-sar-fixtures.json", import.meta.url)));
const keyRegistry = JSON.parse(fs.readFileSync(new URL("./fixtures/portable-sar-fixture-keys.json", import.meta.url)));

function b64urlToBytes(str) {
  const pad = (4 - (str.length % 4)) % 4;
  const b64 = str.replace(/-/g, "+").replace(/_/g, "/") + "=".repeat(pad);
  return new Uint8Array(Buffer.from(b64, "base64"));
}

async function keyPolicy(kid) {
  const entry = keyRegistry[kid];
  if (!entry) return null;
  return { pubkey: b64urlToBytes(entry.pubkey_b64url), profiles: new Set(entry.profiles), source: entry.source };
}

const expected = {
  "01_valid_portable": STATUS.VERIFIED,
  "02_unsigned_counterparty": STATUS.VERIFIED,
  "03_with_ext": STATUS.VERIFIED,
  "04_tampered_signed_field": STATUS.RECEIPT_ID_MISMATCH,
  "05_substituted_unsigned_metadata": STATUS.VERIFIED,
  "05b_stripped_unsigned_metadata": STATUS.VERIFIED,
  "06_unknown_key": STATUS.UNKNOWN_KEY,
  "07_spoofed_key": STATUS.INVALID_SIGNATURE,
  "08_invalid_signature": STATUS.INVALID_SIGNATURE,
  "09_malformed_version": STATUS.NOT_CANDIDATE,
  "09b_malformed_alg": STATUS.NOT_CANDIDATE,
  "11_walletbound_presented_as_portable": STATUS.NOT_CANDIDATE,
  "12_metadata_grafted_walletbound": STATUS.PROFILE_NOT_AUTHORIZED,
  "13_portable_presented_as_walletbound_boundary_case": STATUS.VERIFIED,
  "14_ambiguous_envelope": STATUS.VERIFIED,
};

let pass = 0, fail = 0;
for (const [name, expectedStatus] of Object.entries(expected)) {
  const receipt = fixtures[name];
  const result = await verifyPortableSarReceipt(receipt, keyPolicy);
  const ok = result.status === expectedStatus;
  if (ok) pass++; else fail++;
  console.log(`${ok ? "ok  " : "FAIL"}  ${name.padEnd(45)} expected=${expectedStatus} got=${result.status}${ok ? "" : "  reason=" + result.reason}`);
}

// wallet_binding_attested must never be true for this verifier
for (const [name, expectedStatus] of Object.entries(expected)) {
  if (expectedStatus !== STATUS.VERIFIED) continue;
  const result = await verifyPortableSarReceipt(fixtures[name], keyPolicy);
  const ok = result.walletBindingAttested === false;
  if (ok) pass++; else fail++;
  console.log(`${ok ? "ok  " : "FAIL"}  ${name.padEnd(45)} walletBindingAttested must be false, got=${result.walletBindingAttested}`);
}

// explicit-request-cannot-override-authorization, direct test
{
  const r = await verifyPortableSarReceipt(fixtures["12_metadata_grafted_walletbound"], keyPolicy, "portable-sar-v0.1");
  const ok = r.status === STATUS.PROFILE_NOT_AUTHORIZED;
  if (ok) pass++; else fail++;
  console.log(`${ok ? "ok  " : "FAIL"}  explicit_request_cannot_override_authorization got=${r.status}`);
}

console.log(`\n${pass} passed, ${fail} failed`);
process.exit(fail === 0 ? 0 : 1);
